### PR TITLE
Document Execution Lock

### LINKED ISSUE #473: Do not run multiple pipelines in parallel on the same branch
CirrusCI triggers a new pipeline run on each push, whatever the other runs. As a consequence pushing multiple times in the same branch triggers multiple runs that may overlap. 
That can lead to unexpected side-effects, for example on "stateful" code analysers or on deployments. The penultimate version can be analysed/deployed after the latest version.
I consider this behavior as critical for CI/CD pipelines. Only a single pipeline should run at a point of time in a branch. Trigger requests should be handled in sequence. 

Meanwhile a fix (I hope so!) is implemented, which workaround could be applied?
Thanks

### PR DESCRIPTION
**Pull request by fkorotkov**
Fixes #473

cc @simonbrandhof

### CODE REVIEW

File: docs/guide/writing-tasks.md [hunk 1]
```diff
@@ -649,6 +649,22 @@ task:
 
 You'll be able to manually trigger such paused tasks via Cirrus CI Web UI or directly from GitHub Checks page.
 
+## Task Execution Lock
+
+Some CI tasks perform external operations which are required to be executed one at a time. For example, parallel deploys
+to the same environment is usually a bad idea. In order to restrict parallel execution of a certain task within a repository
```
Comment (RDIL) [inline docs/guide/writing-tasks.md]:
```suggestion
to the same environment is usually a bad idea. In order to restrict parallel execution of a certain task within a repository,
```

File: docs/guide/writing-tasks.md [hunk 2]
```diff
@@ -649,6 +649,22 @@ task:
...
+you can use `execution_lock` to specify a lock key that will be used to make sure that tasks with the same lock key
```
Comment (RDIL) [inline docs/guide/writing-tasks.md]:
```suggestion
you can use `execution_lock` to specify a lock key, a unique string that will be used to make sure that tasks with the same `execution_lock` string
```

File: docs/guide/writing-tasks.md [hunk 3]
```diff
@@ -649,6 +649,22 @@ task:
 
 You'll be able to manually trigger such paused tasks via Cirrus CI Web UI or directly from GitHub Checks page.
 
+## Task Execution Lock
+
+Some CI tasks perform external operations which are required to be executed one at a time. For example, parallel deploys
+to the same environment is usually a bad idea. In order to restrict parallel execution of a certain task within a repository
+you can use `execution_lock` to specify a lock key that will be used to make sure that tasks with the same lock key 
+are executed one at a time. Here is an example of how to make sure that for a single branch there is at most one deployment
```
Comment (RDIL) [inline docs/guide/writing-tasks.md]:
```suggestion
are executed one at a time. Here is an example of how to make sure deployments 
```

File: docs/guide/writing-tasks.md [hunk 4]
```diff
@@ -649,6 +649,22 @@ task:
 
 You'll be able to manually trigger such paused tasks via Cirrus CI Web UI or directly from GitHub Checks page.
 
+## Task Execution Lock
+
+Some CI tasks perform external operations which are required to be executed one at a time. For example, parallel deploys
+to the same environment is usually a bad idea. In order to restrict parallel execution of a certain task within a repository
+you can use `execution_lock` to specify a lock key that will be used to make sure that tasks with the same lock key 
+are executed one at a time. Here is an example of how to make sure that for a single branch there is at most one deployment 
+to staging is happening:
```
Comment (RDIL) [inline docs/guide/writing-tasks.md]:
```suggestion
on a specific branch *can not* run in parallel:
```

File: docs/guide/writing-tasks.md [hunk 5]
```diff
@@ -649,6 +649,22 @@ task:
...
+
+```yaml
+task:
+  name: "Automatic Staging Deploy"
+  execution_lock: $CIRRUS_BRANCH
+```
+
+You'll be able to manually trigger such paused tasks via Cirrus CI Web UI or directly from GitHub Checks page.
```
Comment (RDIL) [inline docs/guide/writing-tasks.md]:
```suggestion
You'll be able to manually trigger such paused tasks via the [Cirrus CI Web Dashboard](https://cirrus-ci.com) or directly from the commit's `checks` page on GitHub.
```

File: docs/guide/writing-tasks.md (hunk 2 above)
Comment (fkorotkov) [reply docs/guide/writing-tasks.md]: That is an awesome clarification! Thanks!

**Pull request merged**